contracts/amm/SirenExchange.sol

```diff
@@ -0,0 +1,280 @@
+            msg.sender,
+            address(this),
+            SeriesLibrary.bTokenIndex(seriesId),
+            bTokenAmount,
+            dataReturn()
+        );
+
+        erc1155Controller.setApprovalForAll(address(sirenAmmAddress), true);
+
+        ISirenTradeAMM(sirenAmmAddress).bTokenSell(
+            seriesId,
+            bTokenAmount,
+            collateralAmountIn
+        );
+
+        TransferHelper.safeApprove(
+            path[0],
+            address(IUniswapV2Router02(_router)),
+            amountsOut[0]
+        );
+        // Executes the swap, returning the amountIn actually spent.
+        amounts = IUniswapV2Router02(_router).swapExactTokensForTokens(
+            collateralAmountIn,
+            amountsOut[amountsOut.length - 1],
+            path,
+            msg.sender,
+            deadline
+        );
+
+        emit BTokenSell(
+            amounts,
+            path,
+            sirenAmmAddress,
+            bTokenAmount,
+            seriesId,
+            msg.sender
+        );
+        return amounts;
+    }
+
+    /// @notice Sell the wToken of a given series to the AMM in exchange for collateral token
+    /// @param seriesId The ID of the Series to buy wToken on
+    /// @param wTokenAmount The amount of wToken to sell (wToken has the same decimals as the underlying)
+    /// @param path The path of the collateral token of the series to the collateral the caller wishes to receive
+    /// @param tokenAmountOutMinimum The lowest amount of collateral the caller is willing to receive as payment
+    /// @param sirenAmmAddress address of the amm that we wish to call
+    /// @param deadline deadline the transaction must be completed by
+    /// @param _router address of the router we wish to use ( QuickSwap or SushiSwap )
+    /// We supply a wToken and then select which collateral we wish to receive as our payment ( if it isnt the underlying asset )
+    function wTokenSell(
+        uint64 seriesId,
+        uint256 wTokenAmount,
+        address[] calldata path,
+        uint256 tokenAmountOutMinimum,
+        address sirenAmmAddress,
+        uint256 deadline,
+        address _router
+    ) external returns (uint256[] memory amounts) {
+        uint256 collateralAmountIn = ISirenTradeAMM(sirenAmmAddress)
+            .wTokenGetCollateralOut(seriesId, wTokenAmount);
+
+        uint256[] memory amountsOut = IUniswapV2Router02(_router).getAmountsOut(
+            collateralAmountIn,
+            path
+        );
+
+        require(
+            amountsOut[amountsOut.length - 1] >= tokenAmountOutMinimum,
+            "Minimum token ammunt out not met"
+        );
+
+        //Transfer token back to the user
+        erc1155Controller.safeTransferFrom(
+            msg.sender,
+            address(this),
+            SeriesLibrary.wTokenIndex(seriesId),
+            wTokenAmount,
+            dataReturn()
+        );
+
+        erc1155Controller.setApprovalForAll(address(sirenAmmAddress), true);
```

Comment:
Random thought... but should we be revoking this "approveForAll" functionality after the trade is executed?

Reply:
I don't see anywhere else in our code that we revoke it however that is a good thought.... I can look into this, @zareth-san  if you have a thought on this that would be appreciated.